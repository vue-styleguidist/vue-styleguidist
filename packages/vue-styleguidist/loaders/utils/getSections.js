// This two functions should be in the same file because of cyclic imports

const fs = require('fs')
const path = require('path')
const _ = require('lodash')
const requireIt = require('react-styleguidist/lib/loaders/utils/requireIt')
const getComponentFiles = require('react-styleguidist/lib/loaders/utils/getComponentFiles')
const slugger = require('react-styleguidist/lib/loaders/utils/slugger')
const getComponents = require('./getComponents')

const examplesLoader = path.resolve(__dirname, '../examples-loader.js')

/**
 * Return object for one level of sections.
 *
 * @param {Array} sections
 * @param {object} config
 * @param {number} parentDepth
 * @returns {Array}
 */
function getSections(sections, config, parentDepth) {
	return sections.map(section => processSection(section, config, parentDepth))
}

const getSectionComponents = (section, config) => {
	let ignore = config.ignore ? _.castArray(config.ignore) : []
	if (section.ignore) {
		ignore = ignore.concat(_.castArray(section.ignore))
	}

	return getComponents(getComponentFiles(section.components, config.configDir, ignore), config)
}

/**
 * Return an object for a given section with all components and subsections.
 * @param {object} section
 * @param {object} config
 * @param {number} parentDepth
 * @returns {object}
 */
function processSection(section, config, parentDepth) {
	const contentRelativePath = section.content

	// Try to load section content file
	let content
	if (contentRelativePath) {
		const contentAbsolutePath = path.resolve(config.configDir, contentRelativePath)
		if (!fs.existsSync(contentAbsolutePath)) {
			throw new Error(`Styleguidist: Section content file not found: ${contentAbsolutePath}`)
		}
		content = requireIt(`!!${examplesLoader}?customLangs=vue|js|jsx!${contentAbsolutePath}`)
	}

	let sectionDepth

	if (parentDepth === undefined) {
		sectionDepth = section.sectionDepth !== undefined ? section.sectionDepth : 0
	} else {
		sectionDepth = parentDepth === 0 ? 0 : parentDepth - 1
	}

	return {
		name: section.name,
		exampleMode: section.exampleMode || config.exampleMode,
		usageMode: section.usageMode || config.usageMode,
		sectionDepth,
		description: section.description,
		slug: slugger.slug(section.name),
		sections: getSections(section.sections || [], config, sectionDepth),
		filepath: contentRelativePath,
		href: section.href,
		components: getSectionComponents(section, config),
		content,
		external: section.external
	}
}

module.exports = getSections
module.exports.processSection = processSection
