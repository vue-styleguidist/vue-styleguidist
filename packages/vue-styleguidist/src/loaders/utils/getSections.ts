// These two functions should be in the same file because of cyclic imports

import * as fs from 'fs'
import * as path from 'path'
import { castArray, flatten } from 'lodash'
import * as Rsg from 'react-styleguidist'
import { parseMulti, ParamTag, ScriptHandlers } from 'vue-docgen-api'
import requireIt from 'react-styleguidist/lib/loaders/utils/requireIt'
import getComponentFiles from 'react-styleguidist/lib/loaders/utils/getComponentFiles'
import slugger from 'react-styleguidist/lib/loaders/utils/slugger'
import { SanitizedStyleguidistConfig } from '../../types/StyleGuide'
import getComponents from './getComponents'

const examplesLoader = path.resolve(__dirname, '../examples-loader.js')

export interface SectionFunctionOptions {
	config: SanitizedStyleguidistConfig
	componentFiles: string[]
	requiredComponents?: Record<string, string[]>
	requiredComponentsList?: string[]
	parentDepth?: number
}

/**
 * Return object for one level of sections.
 *
 * @param {Array} sections
 * @param {object} config
 * @param {number} parentDepth
 * @returns {Array}
 */
export default async function getSections(
	sections: Rsg.ConfigSection[],
	opts: SectionFunctionOptions
): Promise<Rsg.LoaderSection[]> {
	const { config, componentFiles } = opts

	if (!opts.requiredComponents) {
		opts.requiredComponents = await getRequiredComponents(componentFiles, config.jsxInComponents)
		opts.requiredComponentsList = flatten(Object.values(opts.requiredComponents))
	}

	return Promise.all(sections.map(async section => await processSection(section, opts)))
}

/**
 * Returns all the `@required` file path in the analyzed components
 * this way we can ignore them when analyzing the components in the menu
 * and only add them as a subsection for a parent component
 * @param componentFiles all the component file paths to be analyzed
 */
export async function getRequiredComponents(
	componentFiles: string[],
	jsx: boolean
): Promise<Record<string, string[]>> {
	const pathsArrays = await Promise.all(
		componentFiles.map(async componentPath => {
			const compDirName = path.dirname(componentPath)

			try {
				const docs = await parseMulti(componentPath, {
					scriptPreHandlers: [],
					scriptHandlers: [ScriptHandlers.componentHandler],
					jsx
				})
				return docs.reduce(
					(acc: string[], doc) => {
						if (doc.tags && doc.tags.requires) {
							acc = acc.concat(
								doc.tags.requires.map((t: ParamTag) =>
									path.resolve(compDirName, t.description as string)
								)
							)
						}
						return acc
					},
					[componentPath]
				)
			} catch (e) {
				// eat the error (reported in vuedoc-loader)
			}
			return []
		})
	)
	return pathsArrays.reduce((acc: Record<string, string[]>, pa) => {
		const parentComponent = (pa && pa[0]) || 'undefined'
		acc[parentComponent] = (pa && pa.slice(1)) || []
		return acc
	}, {})
}

/**
 * Return an object for a given section with all components and subsections.
 * @param {object} section
 * @param {object} config
 * @param {number} parentDepth
 * @returns {object}
 */
export async function processSection(
	section: Rsg.ConfigSection,
	opts: SectionFunctionOptions
): Promise<Rsg.LoaderSection> {
	const { config, parentDepth } = opts
	const contentRelativePath = section.content

	// Try to load section content file
	let content
	if (contentRelativePath) {
		const contentAbsolutePath = path.resolve(config.configDir, contentRelativePath)
		if (!fs.existsSync(contentAbsolutePath)) {
			throw new Error(`Styleguidist: Section content file not found: ${contentAbsolutePath}`)
		}
		content = requireIt(`!!${examplesLoader}?customLangs=vue|js|jsx!${contentAbsolutePath}`)
	}

	let sectionDepth

	if (parentDepth === undefined) {
		sectionDepth = section.sectionDepth !== undefined ? section.sectionDepth : 0
	} else {
		sectionDepth = parentDepth === 0 ? 0 : parentDepth - 1
	}

	return {
		name: section.name || '',
		exampleMode: section.exampleMode || config.exampleMode,
		usageMode: section.usageMode || config.usageMode,
		sectionDepth,
		description: section.description,
		slug: slugger.slug(section.name || ''),
		sections: await getSections(section.sections || [], { ...opts, parentDepth: sectionDepth }),
		href: section.href,
		components: getSectionComponents(section, opts),
		content,
		external: section.external
	}
}

const getSectionComponents = (
	section: Rsg.ConfigSection,
	opts: SectionFunctionOptions
): Rsg.LoaderComponent[] => {
	const { config, requiredComponentsList, requiredComponents } = opts
	let ignore = config.ignore ? castArray(config.ignore) : []

	if (section.ignore) {
		ignore = ignore.concat(castArray(section.ignore))
	}

	if (requiredComponentsList) {
		ignore = ignore.concat(requiredComponentsList)
	}

	return getComponents(
		getComponentFiles(section.components, config.configDir, ignore),
		config,
		requiredComponents
	)
}
