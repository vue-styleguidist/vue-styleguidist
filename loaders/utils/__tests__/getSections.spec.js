import path from 'path';
import getSections, { processSection } from '../getSections';

const configDir = path.resolve(__dirname, '../../../test');
const sections = [
	{
		name: 'Readme',
		content: 'components/Button/Readme.md',
	},
	{
		name: 'Components',
		components: 'components/**/[A-Z]*.js',
	},
	{
		name: 'Ignore',
		components: 'components/**/*.js',
		ignore: '**/components/Annotation/*',
	},
];
const config = {
	configDir,
	getExampleFilename: a => a,
	getComponentPathLine: a => a,
};
describe('processSection', () => {
	it('should return an object for section with content', () => {
		const result = processSection(sections[0], config);

		expect(result).toMatchSnapshot();
	});

	it('should throw when content file not found', () => {
		const fn = () => processSection({ content: 'pizza' }, config);

		expect(fn).toThrowError('Section content file not found');
	});

	xit('should return an object for section with components', () => {
		const result = processSection(sections[1], config);

		expect(result).toMatchSnapshot();
	});

	xit('should return an object for section without ignored components', () => {
		const result = processSection(sections[2], config);

		expect(result).toMatchSnapshot();
	});
});

xdescribe('getSections', () => {
	it('should return an array', () => {
		const result = getSections(sections, config);

		expect(result).toMatchSnapshot();
	});
});
