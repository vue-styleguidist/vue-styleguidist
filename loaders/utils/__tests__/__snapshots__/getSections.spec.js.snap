// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processSection should return an object for section with content 1`] = `
Object {
  "components": Array [],
  "content": Object {
    "require": "!!~/loaders/examples-loader.js!~/test/components/Button/Readme.md",
  },
  "description": undefined,
  "filepath": "components/Button/Readme.md",
  "name": "Readme",
  "sections": Array [],
  "slug": "readme",
}
`;
